src/net.cpp

```diff
@@ -462,19 +462,22 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
     if (pszDest) {
         const std::vector<CService> resolved{Lookup(pszDest, default_port, fNameLookup && !HaveNameProxy(), 256)};
         if (!resolved.empty()) {
-            const CService& rnd{resolved[GetRand(resolved.size())]};
-            addrConnect = CAddress{MaybeFlipIPv6toCJDNS(rnd), NODE_NONE};
-            if (!addrConnect.IsValid()) {
-                LogPrint(BCLog::NET, "Resolver returned invalid address %s for %s\n", addrConnect.ToStringAddrPort(), pszDest);
-                return nullptr;
-            }
-            // It is possible that we already have a connection to the IP/port pszDest resolved to.
-            // In that case, drop the connection that was just created.
-            LOCK(m_nodes_mutex);
-            CNode* pnode = FindNode(static_cast<CService>(addrConnect));
-            if (pnode) {
-                LogPrintf("Failed to open new connection, already connected\n");
-                return nullptr;
+            // If the connection is made by name, it can be the case that the name resolves to more than one address.
+            // We don't want to connect any more of them if we are already connected to one
+            for (auto &r : resolved) {
+                addrConnect = CAddress{MaybeFlipIPv6toCJDNS(r), NODE_NONE};
+                if (!addrConnect.IsValid()) {
+                    LogPrint(BCLog::NET, "Resolver returned invalid address %s for %s\n", addrConnect.ToStringAddrPort(), pszDest);
+                    return nullptr;
+                }
+                // It is possible that we already have a connection to the IP/port pszDest resolved to.
+                // In that case, drop the connection that was just created.
+                LOCK(m_nodes_mutex);
+                CNode* pnode = FindNode(static_cast<CService>(addrConnect));
+                if (pnode) {
+                    LogPrintf("Failed to open new connection, already connected\n");
```

Comment:
This is now:

```cpp
LogPrintf("Not opening a connection to %d, already connected to %s\n", pnode->addr.ToStringAddrPort(), addrConnect..ToStringAddrPort());
```

It will print two IP addresses, which are most likely equal. It would be more valuable to print the hostname: "... pool.bitcoinnodes.org, already connected to 1.2.3.4" which implies that 1.2.3.4 is the same as pool.bitcoinnodes.org.

```cpp
LogPrintf("Not opening a connection to %s, already connected to %s\n", pszDest, addrConnect..ToStringAddrPort());
```

Reply:
Addressed in [079c56c](https://github.com/bitcoin/bitcoin/pull/28155/commits/079c56cfe7ed150a158b619f04e6d47edd01ca15)